### PR TITLE
add SetTruncation

### PR DESCRIPTION
**Pull request by riaqn**
I'm not sure if this can be simplied further.

### CODE REVIEW

**Comment by Saizan** [comment]
Thanks for the PR!

We should just think of a better name than `someEquiv`, maybe `universalEquiv` as that's another formulation of the universal property?

File: Cubical/HITs/SetTruncation.agda [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{-
+
+This file contains:
+
+- Definition of set truncation and its eliminator
+
+It should *not* depend on the Agda standard library
```
Comment (mortberg) [inline Cubical/HITs/SetTruncation.agda]: I think we can remove this comment from the header

File: Cubical/HITs/SetTruncation.agda [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    A : Set ℓ
+
+-- lemma 6.9.1 in HoTT book
+elimSetTrunc : {B : ∥ A ∥₀ → Set ℓ} →
+               ((x : ∥ A ∥₀) → isSet (B x)) →
+               (g : (a : A) → B (∣ a ∣₀)) →
+               (x : ∥ A ∥₀) → B x
+elimSetTrunc Bset g ∣ a ∣₀ = g a 
+elimSetTrunc {A = A} {B = B} Bset g (squash₀ x y p q i j) =
+  BsetOver (squash₀ x y p q) (elimSetTrunc Bset g x) (elimSetTrunc Bset g y) (cong (elimSetTrunc Bset g) p) (cong (elimSetTrunc Bset g) q) i j
```
Comment (mortberg) [inline Cubical/HITs/SetTruncation.agda]: This line and some others are very long. Can you please try to make them a bit shorter?

File: Cubical/HITs/SetTruncation.agda [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    where
+    BsetOver' : {x y : ∥ A ∥₀} → (p : x ≡ y) → ∀ bx by bp bp' → bp ≡ bp'
+    BsetOver' {x = x} = J (λ y (p : x ≡ y) → ∀ bx by → (bp bp' : PathP (λ i → B (p i)) bx by) → bp ≡ bp') (Bset x)
+
+    BsetOver : {x y : ∥ A ∥₀} → {p q : x ≡ y} (sq : p ≡ q) → ∀ bx by bp bq → 
+               PathP (λ i → PathP (λ j → B (sq i j)) bx by) bp bq
+    BsetOver {x = x} {y = y} {p = p} = J (λ q (sq : p ≡ q) → ∀ bx by bp bq → PathP (λ i → PathP (λ j → B (sq i j)) bx by) bp bq) (BsetOver' p)
+
+
+someEquiv : {B : Set ℓ} → (isSet B) → (∥ A ∥₀ → B) ≃ (A → B)
```
Comment (mortberg) [inline Cubical/HITs/SetTruncation.agda]: A better name would be good. Maybe `setTruncUniversal`?

File: Cubical/HITs/SetTruncation.agda [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+someEquiv {B = B} Bset = isoToEquiv intro elim (λ y → funExt (λ a → refl)) rightInv
+  where
+  intro = (λ h a → h ∣ a ∣₀) 
+  elim = elimSetTrunc {B = λ x → B} (λ x → Bset)
+
+  rightInv : ∀ h →  elim (intro h) ≡ h
+  rightInv h = funExt (λ x → elimSetTrunc {B = λ x → elim (intro h) x ≡ h x} (λ x → isProp→isSet (Bset (elim (intro h) x) (h x))) (λ a → refl) x)
```
Comment (mortberg) [inline Cubical/HITs/SetTruncation.agda]: Maybe the B in `elimSetTrunc` shouldn't be implicit as you have to give it explicitly all the time?

Also, I think this proof would be a bit slicker if we inline funExt.

**Comment by riaqn** [comment]
Hi Anders, I've modified the code per your suggestions.

File: Cubical/HITs/SetTruncation.agda (hunk 1 above)
Comment (riaqn) [reply Cubical/HITs/SetTruncation.agda]: fixed.

File: Cubical/HITs/SetTruncation.agda (hunk 2 above)
Comment (riaqn) [reply Cubical/HITs/SetTruncation.agda]: fixed.

File: Cubical/HITs/SetTruncation.agda (hunk 3 above)
Comment (riaqn) [reply Cubical/HITs/SetTruncation.agda]: fixed.

File: Cubical/HITs/SetTruncation.agda (hunk 4 above)
Comment (riaqn) [reply Cubical/HITs/SetTruncation.agda]: I remove all the explicit B.

**Comment by mortberg** [comment]
@riaqn : are you planning on adding any of the things we discussed today to this PR or should I merge it?

**Comment by mortberg** [comment]
Actually this should go in a separate folder in the same way as the other HITs are defined. It should then also be added to the Everything file

**Pull request merged**